app/models/user.rb

```diff
@@ -46,4 +47,7 @@ def authenticatable_salt
     self.read_attribute(Devise.saml_session_index_key.to_sym) if self.read_attribute(Devise.saml_session_index_key.to_sym).present?
   end
 
+  def get_role_by_grant(grant:)
+    self.grant_permission_role = self.system_admin? ? SYSTEM_ADMIN_GRANT_ROLE : grant.grant_permissions.find_by(user: self).try(:role)
```

Comment:
Consider making system admin role be controlled by the rest of the roles machinery.  Instead of a boolean.

Reply:
Something like this in the GrantPermission model?

`  def self.role_by_user_and_grant(user:, grant:)
    return GrantPermission::ROLES[:admin] if user.system_admin?
    GrantPermission.find_by(grant: grant, user: user)&.role
  end`

Reply:
Where this method would change to something like:

`self.grant_permission_role || GrantPermission.role_by_user_and_grant(user: user, grant: grant)`

Reply:
Yes, that seems right.